pkg/cbfs/file.go

```diff
@@ -0,0 +1,190 @@
+	}
+	if string(f.Magic[:]) != FileMagic {
+		return nil, CbfsHeaderMagicNotFound
+	}
+	Debug("Found CBFS file at %#02x is %v type %v", f.RecordStart, f, f.Type)
+
+	var nameSize uint32
+	if f.AttrOffset == 0 {
+		nameSize = f.SubHeaderOffset - uint32(binary.Size(FileHeader{}))
+	} else {
+		nameSize = f.AttrOffset - uint32(binary.Size(FileHeader{}))
+	}
+	if err := ReadName(r, &f, nameSize); err != nil {
+		return nil, err
+	}
+	if err := ReadAttributes(r, &f); err != nil {
+		return nil, err
+	}
+	if err := ReadData(r, &f); err != nil {
+		return nil, err
+	}
+
+	return &f, nil
+}
+
+// ReadName reads the variable length CBFS name.
+func ReadName(r io.Reader, f *File, size uint32) error {
+	b := make([]byte, size)
+	n, err := r.Read(b)
+	if err != nil {
+		Debug("ReadName failed:%v", err)
+		return err
+	}
+	fname := cleanString(string(b))
+	Debug("ReadName gets '%s' (%#02x)", fname, b)
+	if n != len(b) {
+		err = fmt.Errorf("ReadName: got %d, want %d for name", n, len(b))
+		Debug("ReadName short: %v", err)
+		return err
+	}
+	// discard trailing NULLs
+	z := bytes.Split(b, []byte{0})
+	Debug("ReadName stripped: '%s'", z)
+	f.Name = string(z[0])
+	return nil
+}
+
+// ReadAttributes reads the variable length CBFS attribute list.
+func ReadAttributes(r io.Reader, f *File) error {
+	if f.AttrOffset == 0 {
+		return nil
+	}
+
+	b := make([]byte, f.SubHeaderOffset-f.AttrOffset)
+	n, err := r.Read(b)
+	if err != nil {
+		Debug("ReadAttributes failed:%v", err)
+		return err
+	}
+	Debug("ReadAttributes gets %#02x", b)
+	if n != len(b) {
+		err = fmt.Errorf("ReadAttributes: got %d, want %d for name", n, len(b))
+		Debug("ReadAttributes short: %v", err)
+		return err
+	}
+	f.Attr = b
+	return nil
+}
+
+// ReadData reads the variable length CBFS file data.
+func ReadData(r io.ReadSeeker, f *File) error {
+	Debug("ReadData: Seek to %#x", int64(f.RecordStart+f.SubHeaderOffset))
+	if _, err := r.Seek(int64(f.RecordStart+f.SubHeaderOffset), io.SeekStart); err != nil {
+		return err
+	}
+	Debug("ReadData: read %#x", f.Size)
+	b := make([]byte, f.Size)
+	n, err := r.Read(b)
+	if err != nil {
+		Debug("ReadData failed:%v", err)
+		return err
+	}
+	f.FData = b
+	Debug("ReadData gets %#02x", n)
+	return nil
+}
+
+// FindAttribute returns the attribute with given tag as
+// []byte. It has the size as specified by the tag.
+// Returns an error if not found or could not read in total.
+func (f *File) FindAttribute(t Tag) ([]byte, error) {
+	buf := bytes.NewReader(f.Attr)
+	generic := FileAttr{}
+
+	for {
+		if err := binary.Read(buf, Endian, &generic); err != nil {
+			return nil, err
+		}
+
+		Debug("FindAttribute: Found attribute with tag %x", generic.Tag)
+
+		if Tag(generic.Tag) == t {
+			_, _ = buf.Seek(-int64(binary.Size(generic)), io.SeekCurrent)
+
+			ret := make([]byte, generic.Size)
+			err := binary.Read(buf, Endian, &ret)
+			return ret, err
+		} else {
+			_, err := buf.Seek(int64(generic.Size)-int64(binary.Size(generic)), io.SeekCurrent)
```

Comment:
This is where the issue arises: When encountering a `0`, `generic.Size` is `0` and `binary.Size(generic)` is `-8`, so this will seek by `-8`. Split the calculation out, and ensure to seek by 1 at least:

```go
seek := int64(generic.Size)-int64(binary.Size(generic))
if seek <= 0 {
  seek = 1
}
_, err := buf.Seek(seek, io.SeekCurrent)
```